src/ray/object_manager/object_directory.cc

```diff
@@ -8,30 +8,30 @@ ObjectDirectory::ObjectDirectory(boost::asio::io_service &io_service,
 
 namespace {
 
-/// Process a suffix of the object table log and store the result in
+/// Process a notification of the object table entries and store the result in
 /// client_ids. This assumes that client_ids already contains the result of the
-/// object table log up to but not including this suffix. This also stores a
+/// object table entries up to but not including this notification. This also stores a
 /// bool in has_been_created indicating whether the object has ever been
 /// created before.
-void UpdateObjectLocations(const std::vector<ObjectTableDataT> &location_history,
+void UpdateObjectLocations(const GcsTableNotificationMode mode,
+                           const std::vector<ObjectTableDataT> &location_updates,
                            const ray::gcs::ClientTable &client_table,
                            std::unordered_set<ClientID> *client_ids,
                            bool *has_been_created) {
-  // location_history contains the history of locations of the object (it is a log),
-  // which might look like the following:
-  //   client1.is_eviction = false
-  //   client1.is_eviction = true
-  //   client2.is_eviction = false
-  // In such a scenario, we want to indicate client2 is the only client that contains
-  // the object, which the following code achieves.
-  if (!location_history.empty()) {
+  // location_updates contains the updates of locations of the object.
+  // with GcsTableNotificationMode, we can determine whether the update mode is
+  // addition or deletion (or just current value).
+  if (!location_updates.empty()) {
     // If there are entries, then the object has been created. Once this flag
     // is set to true, it should never go back to false.
     *has_been_created = true;
```

Comment:
It seems like the `has_been_created` flag isn't really necessary anymore, since the only place we were really using it was for evicted actor objects. We are using it [here](https://github.com/ray-project/ray/blob/master/src/ray/object_manager/object_directory.cc#L207), though, to make sure that we received a first subscription notification before using the cached object locations.

How about we rename `has_been_created` to `subscribed`, and just set it to `true` as soon as we receive the first notification callback for an object? We should also remove it from the `ObjectDirectory` public API.

Reply:
I agree with renaming `has_been_created` to `subscribed`, and I tried. But I found another usage of `has_been_created` [here](https://github.com/ray-project/ray/blob/ff5e3384ce8a662e81a5d54536a278b5aadcadaa/src/ray/raylet/reconstruction_policy.cc#L96). I don't know how should I update it.

Reply:
Hmm if you follow that flag up to the [node manager](https://github.com/ray-project/ray/blob/master/src/ray/raylet/node_manager.cc#L64), it looks like it's not actually used. Sorry, I think this was my bad :)

I think it's okay to just remove the usage in the `ReconstructionPolicy` as well.

Reply:
Done.